text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,266 @@
+
+However, rendering on the server can itself introduce performance issues, working against the original intent. Namely, if DOM that was generated from SSR HTML is not preserved during the initial CSR, this will result in potentially expensive DOM manipulation and unwanted visual artifacts.
+
+The aim of rehydration is to avoid the client-side performance issues of a naive SSR implementation, reusing original DOM wherever possible.
+
+### The happy path
+
+With regard to SSR, there exists a happy path (Fig 1a) where the rendering output for a given request is identical on both the server and the client; identical VDOM is generated for the entire document in both environments.
+
+Depending upon the actual implementation, exceptions to this happy path include:
+
+- **Exception A:** One or more components are intentionally rendered differently on the server and on the client (Fig 1b). For example, a placeholder may be rendered on the server and later replaced on the client.
+  - **Example:** The page contains a cart icon with a dynamic number of items in cart.
+  - **Example:** The page contains a user avatar.
+- **Exception B:** One or more components are rendered only on the server (Fig 1c). No JS is sent to the client for these components and the original SSR'd HTML remains untouched by hydration and later client-side renders.
+  - **Example:** The page contains a footer that never changes.
+  - **Example:** The landing page is mostly static and is cached on a CDN.
+- **Exception C:** Render-critical state is changed between the time of SSR and CSR rehydration, causing an unexpected mismatch (Fig 1d).
+  - **Example:** A product price is updated after SSR but before rehydration on the client.
+  - **Example:** The CDN delivers a stale version of the landing page, which was updated after the SSR HTML was en route to the client.
+
+<figure>
+  <img src="../assets/0000-ssr-rehydration-fig1.png" alt="Fig. 1">
+  <figcaption align = "center">
+    <b>Fig.1: the happy path and its exceptions</b>
+  </figcaption>
+</figure>
+
+## Design
+
+Several pieces need to come together to unlock rehydration. The general flow is illustrated below in figure 2.
+
+<figure>
+  <img src="../assets/0000-ssr-rehydration-fig2.png" alt="Fig. 2" width="60%">
+  <figcaption align = "center">
+    <b>Fig.2: the proposed implementation</b>
+  </figcaption>
+</figure><br /><br />
+
+In roughly chronological order with respect to a single page request, the following steps shall be taken as part of rehydration:
+
+- As part of SSR, serialize render-critical state and bake it into the HTML response returned from the server. For each `LightningElement`, attach corresponding state to its root DOM node as `data-ssr-state`.

```

Comment:
What is `state`? What will be serializable? How it will look like?

Reply:
Where that data-ssr-state comes from? Is it user-land driven? Or is the compiler producing it?